editor/gdextension_reload_node.cpp

```diff
@@ -0,0 +1,79 @@
+/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
+/*                                                                        */
+/* Permission is hereby granted, free of charge, to any person obtaining  */
+/* a copy of this software and associated documentation files (the        */
+/* "Software"), to deal in the Software without restriction, including    */
+/* without limitation the rights to use, copy, modify, merge, publish,    */
+/* distribute, sublicense, and/or sell copies of the Software, and to     */
+/* permit persons to whom the Software is furnished to do so, subject to  */
+/* the following conditions:                                              */
+/*                                                                        */
+/* The above copyright notice and this permission notice shall be         */
+/* included in all copies or substantial portions of the Software.        */
+/*                                                                        */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
+/**************************************************************************/
+
+#include "gdextension_reload_node.h"
+
+#include "core/extension/gdextension_manager.h"
+#include "editor/editor_inspector.h"
+#include "editor/editor_interface.h"
+
+void GDExtensionReloadNode::_bind_methods() {
+}
+
+void GDExtensionReloadNode::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_APPLICATION_FOCUS_OUT: {
+			unfocused = true;
+		} break;
+
+		case NOTIFICATION_APPLICATION_FOCUS_IN: {
+			callable_mp(this, &GDExtensionReloadNode::_deferred_reload).call_deferred();
+		} break;
+	}
+}
+
+void GDExtensionReloadNode::_deferred_reload() {
+	if (!unfocused) {
+		return;
+	}
+	unfocused = false;
+
+	GDExtensionManager *gdextension_manager = GDExtensionManager::get_singleton();
+
+	Vector<String> extensions = gdextension_manager->get_loaded_extensions();
+	bool reloaded = false;
+	for (const String &extension_name : extensions) {
+		Ref<GDExtension> extension = gdextension_manager->get_extension(extension_name);
+
+		if (!extension->is_reloadable()) {
+			continue;
+		}
+
+		if (extension->has_library_changed()) {
+			reloaded = true;
+			gdextension_manager->reload_extension(extension->get_path());
+		}
+	}
+
+	if (reloaded) {
+		// In case the developer is inspecting a node whose class was reloaded.
+		EditorInterface::get_singleton()->get_inspector()->update_tree();
```

Comment:
Updating the tree is pretty expensive. Is there no more reliable way to only update it if the object is actually being inspected? Like a signal from the object itself, similar to `notify_property_list_changed`?

Reply:
Hm, I don't think so.

We'd need to do a whole lot more tracking to even know if something has changed on an extension class, since there's just so many ways a class could change between reloads. (The class itself can't keep track of it, because it's actually two different code bases: the old version of the class was loaded from an older version of the DLL, and the new version is created from a newer version of the DLL, and we have to destroy the old one before we can load the new one.)

I think this probably the best we can do.

But! This is something that only happens when (1) the developer has recompiled the DLL (we check if the modification time has changed) and (2) they have unfocused and refocused the editor. So, it's not something that's happening very often!

Reply:
> For the record, we don't necessarily need to detect specific changes, just that the inspected object is relevant here to trigger an update.

Even that's a little tricky. Like, you could check that the object being inspected is from an extension. But an extension could also provide a `Resource`, and if that resource is used as a value on a property of a non-extension object (especially if the resources properties are expanded), you'd also want to update the inspector tree.

Reply:
Right. By the way, what about the other way around? What if the inspected object does belong to the extension, but after its reloaded this type is no longer available?

Reply:
If an extension is removed between reloads, it changes type to be whatever it's parent class was. So, if the extension class descended from `Node2D`, after the reload, it just becomes a plain `Node2D`.